src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -24605,6 +24605,1499 @@ void Init()
+    public string? Value { get; }
+
+    public static string M(Test t)
+    {
+        if (t.IsOk is true)
+        {
+            return t.Value;
+        }
+        else
+        {
+            return t.Value; // 1
+        }
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics(
+                // (19,20): warning CS8603: Possible null reference return.
+                //             return t.Value; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "t.Value").WithLocation(19, 20)
+                );
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_IsFalseTest()
+        {
+            var c = CreateCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public struct Test
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk { get; }
+    public string? Value { get; }
+
+    public string M()
+    {
+        if (IsOk is false)
+        {
+            return Value; // 1
+        }
+        else
+        {
+            return Value;
+        }
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics(
+                // (15,20): warning CS8603: Possible null reference return.
+                //             return Value; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "Value").WithLocation(15, 20)
+                );
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_SwitchStatementOnProperty()
+        {
+            var c = CreateCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public struct Test
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk { get; }
+    public string? Value { get; }
+
+    public string M()
+    {
+        switch (IsOk)
+        {
+            case true:
+                return Value;
+            default:
+                return Value; // 1
+        }
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics(
+                // (18,24): warning CS8603: Possible null reference return.
+                //                 return Value; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "Value").WithLocation(18, 24)
+                );
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_SwitchExpressionOnProperty()
+        {
+            var c = CreateCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public struct Test
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk { get; }
+    public string? Value { get; }
+
+    public string M()
+    {
+        return IsOk switch
+        {
+            true => Value,
+            _ => throw null!,
```

Comment:
>_ => throw null!, [](start = 12, length = 17)

Consider testing `_ => Value` instead.

Reply:
This results in a diagnostic on the `return`, which doesn't tell us which branch caused the problem:
```
                 // (13,16): warning CS8603: Possible null reference return.
                    //         return IsOk switch
                    Diagnostic(ErrorCode.WRN_NullReferenceReturn, @"IsOk switch
            {
                true => Value,
                _ => Value, // 1
            }").WithLocation(13, 16)
```

---
In reply to: [573115096](https://github.com/dotnet/roslyn/pull/51001#discussion_r573115096) [](ancestors = 573115096)